### ISSUE TITLE
FreeBSD13 reliably crashes on ZFS destroy activity (panic: VERIFY(avl_find(tree, new_node, &where) == NULL) failed)

### ISSUE THREAD

**Issue by bretton** [openzfs/zfs#12559]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| FreeBSD13
Distribution Version	| 13.0-RELEASE-p4
Kernel Version	| 13.0-RELEASE-p4
Architecture	| AMD64
OpenZFS Version	|zfs-2.0.0-FreeBSD_gf11b09dec, zfs-kmod-2.0.0-FreeBSD_gf11b09dec

### Describe the problem you're observing
This may be similar to #11480 and #11603 

Initially ZFS datasets were setup with compression and dedup enabled. However repeat ZFS destroy activity (from `pot destroy` or `zfs destroy`) would trigger a system crash.

Initial diagnosis revealed problems with compression and dedup so these were disabled. ZFS dataset was recreated from scratch and was stable for 24h. 

Test with `compression=lz4` as option worked fine for a bit, lots of ZFS create-and-destroy testing, then turned off.

I thought that would be it, and things would be stable, yet was wrong. After a day's worth of ZFS create/destroy activity, the system started crashing again on ZFS destroy. (I'm building and destroying environments with pot)

As soon as the problem starts, it never goes away. Any attempt to change ZFS on the dataset triggers a crash. The zpool needs to be removed and setup from scratch, which may involve losing data.

It all works on FreeBSD12_2, and is rock solid, even with compression and dedup, but migrating to FreeBSD13_0 results in an unstable, reduced-feature-set system that eventually crashes on a random ZFS destroy command.

Disks in pool been thoroughly checked and no problems.

What can I check? What can I do?

### Describe how to reproduce the problem
ZFS create/destroy multiple times on system with nomad-pot-driver.

### Include any warning/errors/backtraces from the system logs
Crashinfo
```
psf00002 dumped core - see /var/crash/vmcore.8

Sat Sep 11 21:39:20 SAST 2021

FreeBSD psf00002 13.0-RELEASE-p4 FreeBSD 13.0-RELEASE-p4 #0: Tue Aug 24 07:33:27 UTC 2021     root@amd64-builder.daemonology.net:/usr/obj/usr/src/amd64.amd64/sys/GENERIC  amd64

panic: VERIFY(avl_find(tree, new_node, &where) == NULL) failed

GNU gdb (GDB) 10.2 [GDB v10.2 for FreeBSD]
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-portbld-freebsd13.0".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /boot/kernel/kernel...
Reading symbols from /usr/lib/debug//boot/kernel/kernel.debug...

Unread portion of the kernel message buffer:
panic: VERIFY(avl_find(tree, new_node, &where) == NULL) failed

cpuid = 2
time = 1631388101
KDB: stack backtrace:
#0 0xffffffff80c574c5 at kdb_backtrace+0x65
#1 0xffffffff80c09ea1 at vpanic+0x181
#2 0xffffffff8217207a at spl_panic+0x3a
#3 0xffffffff8214a686 at avl_add+0x156
#4 0xffffffff82202020 at dsl_livelist_iterate+0xb0
#5 0xffffffff821c12cf at bpobj_iterate_blkptrs+0xff
#6 0xffffffff821c0d14 at bpobj_iterate_impl+0x174
#7 0xffffffff82201efc at dsl_process_sub_livelist+0x5c
#8 0xffffffff8224d020 at spa_livelist_delete_cb+0x130
#9 0xffffffff823251c7 at zthr_procedure+0xc7
#10 0xffffffff80bc7dde at fork_exit+0x7e
#11 0xffffffff810629de at fork_trampoline+0xe
Uptime: 1h20m20s
Dumping 2771 out of 65159 MB:..1%..11%..21%..31%..41%..51%..61%..71%..81%..91%

__curthread () at /usr/src/sys/amd64/include/pcpu_aux.h:55
55	/usr/src/sys/amd64/include/pcpu_aux.h: No such file or directory.
(kgdb) #0  __curthread () at /usr/src/sys/amd64/include/pcpu_aux.h:55
#1  doadump (textdump=<optimized out>) at /usr/src/sys/kern/kern_shutdown.c:399
#2  0xffffffff80c09a96 in kern_reboot (howto=260)
    at /usr/src/sys/kern/kern_shutdown.c:486
#3  0xffffffff80c09f10 in vpanic (fmt=<optimized out>, ap=<optimized out>)
    at /usr/src/sys/kern/kern_shutdown.c:919
#4  0xffffffff8217207a in spl_panic (file=<optimized out>, 
    func=<optimized out>, line=<unavailable>, fmt=<unavailable>)
    at /usr/src/sys/contrib/openzfs/module/os/freebsd/spl/spl_misc.c:107
#5  0xffffffff8214a686 in avl_add (tree=tree@entry=0xfffffe01b03ca1d8, 
    new_node=new_node@entry=0xfffff80158208c60)
    at /usr/src/sys/contrib/openzfs/module/avl/avl.c:641
#6  0xffffffff82202020 in dsl_livelist_iterate (arg=<optimized out>, 
    bp=0xfffffe021ea23600, bp_freed=1, tx=<optimized out>)
    at /usr/src/sys/contrib/openzfs/module/zfs/dsl_deadlist.c:966
#7  0xffffffff821c12cf in bpobj_iterate_blkptrs (
    bpi=bpi@entry=0xfffff801a926b040, 
    func=func@entry=0xffffffff82201f70 <dsl_livelist_iterate>, 
    arg=<unavailable>, arg@entry=0xfffffe01b03ca200, start=<optimized out>, 
    start@entry=0, tx=tx@entry=0x0, free=free@entry=0)
    at /usr/src/sys/contrib/openzfs/module/zfs/bpobj.c:307
#8  0xffffffff821c0d14 in bpobj_iterate_impl (initial_bpo=<optimized out>, 
    initial_bpo@entry=0xfffff80724aa9220, 
    func=0xffffffff82201f70 <dsl_livelist_iterate>, 
    arg=arg@entry=0xfffffe01b03ca200, tx=tx@entry=0x0, free=<optimized out>, 
    free@entry=0, bpobj_size=<optimized out>, bpobj_size@entry=0x0)
    at /usr/src/sys/contrib/openzfs/module/zfs/bpobj.c:383
#9  0xffffffff821c10f2 in bpobj_iterate_nofree (bpo=<unavailable>, 
    bpo@entry=0xfffff80724aa9220, func=<unavailable>, arg=<unavailable>, 
    arg@entry=0xfffffe01b03ca200, bpobj_size=<unavailable>, 
    bpobj_size@entry=0x0)
    at /usr/src/sys/contrib/openzfs/module/zfs/bpobj.c:541
#10 0xffffffff82201efc in dsl_process_sub_livelist (bpobj=<unavailable>, 
    bpobj@entry=0xfffff80724aa9220, to_free=to_free@entry=0xfffffe01b03ca298, 
    t=t@entry=0xfffff801450fbc00, size=<unavailable>, size@entry=0x0)
    at /usr/src/sys/contrib/openzfs/module/zfs/dsl_deadlist.c:1000
#11 0xffffffff8224d020 in spa_livelist_delete_cb (arg=0xfffffe016bd06000, 
    z=0xfffff801450fbc00) at /usr/src/sys/contrib/openzfs/module/zfs/spa.c:2506
#12 0xffffffff823251c7 in zthr_procedure (arg=arg@entry=0xfffff801450fbc00)
    at /usr/src/sys/contrib/openzfs/module/zfs/zthr.c:241
#13 0xffffffff80bc7dde in fork_exit (
    callout=0xffffffff82325100 <zthr_procedure>, arg=0xfffff801450fbc00, 
    frame=0xfffffe01b03ca480) at /usr/src/sys/kern/kern_fork.c:1069
#14 <signal handler called>
(kgdb) 

```

Diagnostic output
```
# zdb -bcsvL -AAA srv

Traversing all blocks to verify checksums ...

8.30G completed (  22MB/s) estimated time remaining: 0hr 02min 38sec        zdb_blkptr_cb: Got error 97 reading <6046, 0, 5, 0> DVA[0]=<0:1f401fc5000:1000> DVA[1]=<0:1fc015ca000:1000> [L5 DMU dnode] fletcher4 lz4 unencrypted LE contiguous unique double size=20000L/1000P birth=53483L/53483P fill=2 cksum=878d8cff79:1f29ccbe27caf:397a6c46bf10e3c:6e06983245cd22a0 -- skipping
8.91G completed (  22MB/s) estimated time remaining: 0hr 02min 06sec        
Error counts:

	errno  count
	   97  1

	bp count:                751538
	ganged count:                 0
	bp logical:          9982648320      avg:  13282
	bp physical:         9006585344      avg:  11984     compression:   1.11
	bp allocated:        9582747648      avg:  12750     compression:   1.04
	bp deduped:                   0    ref>1:      0   deduplication:   1.00
	Normal class:       12606124032     used:  0.32%

	additional, non-pointer bps of type 0:       1505
	Dittoed blocks on same vdev: 55356

Blocks	LSIZE	PSIZE	ASIZE	  avg	 comp	%Total	Type
     -	    -	    -	    -	    -	    -	     -	unallocated
     2	  32K	   8K	  24K	  12K	 4.00	  0.00	object directory
     3	1.50K	   1K	  24K	   8K	 1.50	  0.00	object array
     1	  16K	   4K	  12K	  12K	 4.00	  0.00	packed nvlist
     -	    -	    -	    -	    -	    -	     -	packed nvlist size
   782	97.8M	20.1M	60.3M	79.0K	 4.86	  0.66	bpobj
     -	    -	    -	    -	    -	    -	     -	bpobj header
     -	    -	    -	    -	    -	    -	     -	SPA space map header
   541	58.3M	22.9M	68.8M	 130K	 2.54	  0.75	SPA space map
    10	 940K	 940K	 940K	  94K	 1.00	  0.01	ZIL intent log
   879	28.8M	3.48M	7.34M	8.55K	 8.28	  0.08	DMU dnode
    30	 120K	 120K	 244K	8.13K	 1.00	  0.00	DMU objset
     -	    -	    -	    -	    -	    -	     -	DSL directory
    29	45.5K	8.50K	  36K	1.24K	 5.35	  0.00	DSL directory child map
    26	  13K	   1K	  24K	  945	13.00	  0.00	DSL dataset snap map
    32	 140K	  32K	  96K	   3K	 4.38	  0.00	DSL props
     -	    -	    -	    -	    -	    -	     -	DSL dataset
     -	    -	    -	    -	    -	    -	     -	ZFS znode
     -	    -	    -	    -	    -	    -	     -	ZFS V0 ACL
 42.2K	3.82G	3.14G	3.16G	76.7K	 1.22	 35.39	ZFS plain file
 1.82K	4.97M	1.56M	5.83M	3.20K	 3.18	  0.06	ZFS directory
    11	  11K	  11K	  88K	   8K	 1.00	  0.00	ZFS master node
     -	    -	    -	    -	    -	    -	     -	ZFS delete queue
  638K	5.09G	5.01G	5.04G	8.10K	 1.02	 56.50	zvol object
     -	    -	    -	    -	    -	    -	     -	zvol prop
     -	    -	    -	    -	    -	    -	     -	other uint8[]
     -	    -	    -	    -	    -	    -	     -	other uint64[]
     -	    -	    -	    -	    -	    -	     -	other ZAP
     -	    -	    -	    -	    -	    -	     -	persistent error log
     8	   1M	 108K	 324K	40.5K	 9.48	  0.00	SPA history
     -	    -	    -	    -	    -	    -	     -	SPA history offsets
     -	    -	    -	    -	    -	    -	     -	Pool properties
     -	    -	    -	    -	    -	    -	     -	DSL permissions
     -	    -	    -	    -	    -	    -	     -	ZFS ACL
     -	    -	    -	    -	    -	    -	     -	ZFS SYSACL
     -	    -	    -	    -	    -	    -	     -	FUID table
     -	    -	    -	    -	    -	    -	     -	FUID table size
    14	8.50K	   2K	  12K	  877	 4.25	  0.00	DSL dataset next clones
     -	    -	    -	    -	    -	    -	     -	scan work queue
    60	35.5K	7.50K	  32K	  546	 4.73	  0.00	ZFS user/group/project used
     -	    -	    -	    -	    -	    -	     -	ZFS user/group/project quota
     -	    -	    -	    -	    -	    -	     -	snapshot refcount tags
 49.7K	 199M	 199M	 596M	  12K	 1.00	  6.53	DDT ZAP algorithm
     2	  32K	   8K	  24K	  12K	 4.00	  0.00	DDT statistics
     -	    -	    -	    -	    -	    -	     -	System attributes
     -	    -	    -	    -	    -	    -	     -	SA master node
    11	16.5K	16.5K	  88K	   8K	 1.00	  0.00	SA attr registration
    22	 352K	  88K	 176K	   8K	 4.00	  0.00	SA attr layouts
     -	    -	    -	    -	    -	    -	     -	scan translations
     -	    -	    -	    -	    -	    -	     -	deduplicated block
     -	    -	    -	    -	    -	    -	     -	DSL deadlist map
     -	    -	    -	    -	    -	    -	     -	DSL deadlist map hdr
     4	3.50K	   3K	  36K	   9K	 1.17	  0.00	DSL dir clones
     -	    -	    -	    -	    -	    -	     -	bpobj subobj
     -	    -	    -	    -	    -	    -	     -	deferred free
     -	    -	    -	    -	    -	    -	     -	dedup ditto
    51	 195K	  25K	  96K	1.88K	 7.80	  0.00	other
  734K	9.30G	8.39G	8.92G	12.5K	 1.11	100.00	Total

Block Size Histogram

  block   psize                lsize                asize
   size   Count   Size   Cum.  Count   Size   Cum.  Count   Size   Cum.
    512:  1.83K   938K   938K  1.83K   938K   938K      0      0      0
     1K:  2.57K  3.35M  4.26M  2.57K  3.35M  4.26M      0      0      0
     2K:  3.86K  9.91M  14.2M  3.86K  9.91M  14.2M      0      0      0
     4K:  53.8K   216M   230M  51.7K   209M   224M  9.67K  38.7M  38.7M
     8K:   639K  4.99G  5.22G   638K  4.99G  5.21G   691K  5.60G  5.63G
    16K:  1.95K  41.6M  5.26G  1.80K  33.6M  5.24G  1.45K  30.3M  5.66G
    32K:  5.27K   258M  5.51G    388  16.5M  5.25G  4.81K   244M  5.90G
    64K:  3.47K   274M  5.78G    296  26.8M  5.28G  4.33K   347M  6.24G
   128K:  20.9K  2.61G  8.39G  32.1K  4.02G  9.30G  21.2K  2.67G  8.91G
   256K:      0      0  8.39G      0      0  9.30G     48  13.0M  8.92G
   512K:      0      0  8.39G      0      0  9.30G      0      0  8.92G
     1M:      0      0  8.39G      0      0  9.30G      0      0  8.92G
     2M:      0      0  8.39G      0      0  9.30G      0      0  8.92G
     4M:      0      0  8.39G      0      0  9.30G      0      0  8.92G
     8M:      0      0  8.39G      0      0  9.30G      0      0  8.92G
    16M:      0      0  8.39G      0      0  9.30G      0      0  8.92G

                            capacity   operations   bandwidth  ---- errors ----
description                used avail  read write  read write  read write cksum
srv                       11.7G 3.61T 2.51K     0 65.9M     0     0     0     0
  mirror                  11.7G 3.61T 2.51K     0 65.9M     0     0     0     0
    /dev/ada4                         1.26K     0 33.0M     0     0     0    20
    /dev/ada5                         1.25K     0 33.0M     0     0     0    20

**Comment by bretton** [comment]
System won't boot now after latest crash. Panic error at login screen

**Comment by bretton** [comment]
Two brief updates:

1. Using these options on zpool creation helps with stability, as covered in the other bug report:

```
zpool create -o feature@livelist=disabled -o feature@zstd_compress=disabled -o
feature@log_spacemap=disabled poolname mirror /dev/ada3 /dev/ada4
```

24 hours in, over 480 ZFS create/clone/destroy operations, and system still stable.

2. Pot is a jail management tool that relies heavily on cloning the zfs datasets. This is why so many ZFS operations required.

https://github.com/pizzamig/pot

**Comment by bretton** [comment]
> ```
> zpool create -o feature@livelist=disabled -o feature@zstd_compress=disabled -o
> feature@log_spacemap=disabled poolname mirror /dev/ada3 /dev/ada4
> ```

Tested with nomad pot images being added & removed. No problems. Related ZFS data sets can be deleted no problem too, so the issue isn't with the nomad pot driver. Crashes only happen with the new default features enabled.

**Comment by behlendorf** [comment]
This issue does look the same as #11480 and it was fixed in PR #12177.  The fix has been backported and will be included in the upcoming 2.0.6 release.

**Comment by bretton** [comment]
> This issue does look the same as #11480 and it was fixed in PR #12177. The fix has been backported and will be included in the upcoming 2.0.6 release.

thanks @behlendorf

**Comment by gmelikov** [comment]
@behlendorf jfyi I didn't found fix in stable releases, only in master.

**Comment by tarkhil** [comment]
Sorry for reopening, looks like I've hit it. Is there any way to recover?

https://forums.freebsd.org/threads/my-zpool-exploded.83497/

**Comment by behlendorf** [comment]
Sorry to hear you got bit by this.  This fix was included in the OpenZFS 2.0.6 release, commit abd0b59e488589159e7fb5305bc876294faf6ce4, so it should make its way in to FreeBSD.
